### PR TITLE
Initial implementation

### PR DESCRIPTION
**Pull request by murilomend**
@gustavoneves12 @joaobrittoneto @geovanemimoso

### CODE REVIEW

**Comment by murilomend** [comment]
[Hyro training exercise: Unit Test](https://trello.com/c/cgfjIxjT/6-hyro-training-exercise-unit-test)

File: package.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+	"name": "signal_generator",
+	"version": "0.0.1",
+
+	"author": "Murilo Mendonca Venancio",
+	"description": "Signal generator for learning basic Hyro concepts",
+
+	"dependencies": {
+		"hyro_api_channel": "*",
+		"hyro_utils": "*",
+		"hyro_msgs": "*"
```
Comment (joaobrittoneto) [inline package.json]: Missing dependencies.
Usually all required packages in CMakeLists.txt in `hyro_find_package(hyro_xxx REQUIRED)` should be in the package.json dependecies

File: src/components/DigitalConverterComponent.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  m_dummy = this->registerOutput<std::vector<int>>("fix_dynamic"_uri, config);
+  
+  /** Setting an output channel for sending the analog signal */
+  m_input  = this->registerInput<Signal>("signals"_uri, config);
+
+  /** Setting an output channel for sending the thresholded signal */
+  m_output = this->registerOutput<float>("digital_signals"_uri, config);
+
+  /** Setting the parameters from configuration */
+  m_amplitude = config.parameters.getParameter<double>("amplitude", 2.0);
```
Comment (joaobrittoneto) [inline src/components/DigitalConverterComponent.cpp]: It seems that `m_amplitude` and `m_threshold` are just used here to initialize the `m_thresh`, is that right?
If it's the case, them can be defined locally, not globally.

Comment (murilomend) [reply src/components/DigitalConverterComponent.cpp]: Actually I am also using it inside the get/set functions of my dynamic properties in order to update the `m_thresh.setThreshold(m_amplitude,m_threshold)` method. Should I split this method into two methods: eg. :  `m_thresh.setAmplitude(m_amplitude)`  and `m_thresh.setThreshold(m_threshold)` ?

Comment (joaobrittoneto) [reply src/components/DigitalConverterComponent.cpp]: Yeah, I'd go for it and avoid storing the same information twice in different places.

```
bool DigitalConverterComponent::setThreshold(double threshold)
{ 
  return m_thresh.setThreshold(m_threshold);
}


double DigitalConverterComponent::getThreshold()
{ 
  return m_thresh.getThreshold();
}
```

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,88 @@
+hyro_init_package(signal_generator VERSION 0.0.1)
+
```
Comment (gustavoneves12) [inline CMakeLists.txt]: @murilomend please, try to avoid too much empty lines.

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+## Specify additional locations of header files
+## Your package locations should be listed before other locations
+
+
+set(SIGNAL_GENERATOR_HEADERS
+  # add all *.h files here
+  include/hyro/SignalGeneratorComponent.hpp
+  include/hyro/DigitalConverterComponent.hpp
+
```
Comment (gustavoneves12) [inline CMakeLists.txt]: Do not need empty lines here. If you want to separate the headers of the components from the headers of the messages, you can add a comment between them.

File: CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  # add all *.cpp files here
+  src/components/SignalGeneratorComponent.cpp
+  src/components/DigitalConverterComponent.cpp
+  src/internal/SignalGenerator.cpp
+  src/internal/Thresholding.cpp
+
+)
+
+
+#qt4_wrap_cpp( ${PROJECT_NAME}_MOC_CPP ${hyro_widgets_HEADERS_MOC} )
```
Comment (gustavoneves12) [inline CMakeLists.txt]: Remove QT stuff from here.

File: include/hyro/DigitalConverterComponent.hpp [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  bool   setThreshold(double threshold);
+  double getThreshold();
+  ///@}
+  
+  
+  Thresholding  m_thresh;
+  double        m_amplitude;
+  double        m_threshold;
+  
+  static std::shared_ptr<hyro::HyroLogger>    s_logger;
```
Comment (gustavoneves12) [inline include/hyro/DigitalConverterComponent.hpp]: I suggest you declare the static variables in a different group of the member variables. I usually group the static variables at the end of the class declaration.

**Comment by gustavoneves12** [comment]
Please, delete the file  `initial_implementation.conf`.

File: src/components/DigitalConverterComponent.cpp (hunk 2 above)
Comment (gustavoneves12) [reply src/components/DigitalConverterComponent.cpp]: :+1:

File: src/components/SignalGeneratorComponent.cpp [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+#include <hyro/SignalGeneratorComponent.hpp>
+
+namespace hyro
+{
+
+std::shared_ptr<hyro::HyroLogger> SignalGeneratorComponent::s_logger = hyro::HyroLoggerManager::CreateLogger("SignalGeneratorComponent");
+
+SignalGeneratorComponent::SignalGeneratorComponent (URI uri): hyro::Component(uri)
+{
+  /** Initialize object here */
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: Your Constructor is doing nothing. I suggest you remove it from here and declare 
```c++
using Component::Component`
```
in class declaration.

File: src/components/SignalGeneratorComponent.cpp [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+  m_signal_generator.setSignalGenerator(m_amplitude,m_frequency,m_cosine);
+  return hyro::Result::RESULT_OK;
+}
+
+
+bool SignalGeneratorComponent::setAmplitude(double amplitude)
+{
+  m_amplitude = amplitude;
+  return m_signal_generator.setSignalGenerator(m_amplitude,m_frequency,m_cosine);
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: In SignalGenerator class I would create setters and getters for the parameters. In this function I would replace this for:
```c++
bool SignalGeneratorComponent::setAmplitude(double amplitude)
{
  m_signal_generator.setAmplitude(ampliture);
  return true;
}

```

File: src/components/SignalGeneratorComponent.cpp [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+double SignalGeneratorComponent::getAmplitude()
+{
+  return this->m_amplitude;
```
Comment (gustavoneves12) [inline src/components/SignalGeneratorComponent.cpp]: With the getter function in `SignalGenerator` class you can do:

```c++
return m_signal_generator.getAmplitude();
```

File: src/examples/signal_generator.cpp [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+  /**
+   * STEP 7: Run Everything
+   */
+  CancellationTokenSource cancellation_token;
+
+  StateMachineSpinner sig_spinner(sig_sm, cancellation_token, 1ms);
+  StateMachineSpinner dig_spinner(dig_sm, cancellation_token, 1ms);
+  
+  /** A thread is created for running only the widgets in a separated context while the following part can be run */
+  std::thread plot_thread([]()
```
Comment (gustavoneves12) [inline src/examples/signal_generator.cpp]: Nice @murilomend! By calling the widget in a separate thread we can set the dynamic parameters using the standard      `std::cin`. However, just for your information, it is recommended execute GUI stuff in the main thread.

File: src/internal/SignalGenerator.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+* @author Murilo Mendonça Venâncio
+* 
+* @date 24/01/2019
+*/
+
+#include "SignalGenerator.hpp"
+
+
+SignalGenerator::SignalGenerator()
+{
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp]: As @joaobrittoneto mentioned before, we usually initialize the parameters as show below:

```
SignalGenerator::SignalGenerator()
  : m_amp(1)
  , m_freq(1)
  , m_cosine(true)
{
}
```

File: src/internal/SignalGenerator.cpp [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+  setSignalGenerator(1,1,true);
+  m_deg  = 0.;
+}
+
+SignalGenerator::~SignalGenerator()
+{
+
+}
+
+SignalGenerator::SignalGenerator(const double &amp,const double &freq,const bool &cosine)
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp]: For native variables (`double`, `int`, `float`, `char`, etc) I would not pass as reference like `const double&`. For vectors and complex data types I would do like you did.

File: src/internal/SignalGenerator.cpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+
+
+double SignalGenerator::getSignalValue()
+{
+  /** Basically this function updates the siusoidal wave by a m_step eachp time. */
+  double rad = m_deg / 180.0 * M_PI;
+  m_deg += m_step;
+  return m_cosine?(m_amp * cos(rad)):(m_amp * sin(rad));
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp:54]: :1st_place_medal: :+1:

File: tests/test_signal_generator.cpp [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+  
+  /** Fake input channels tests */
+  auto inputSignal = std::make_shared<FakeInput<Signal>>("inputSignal"_uri, "api", "/signal_generator/signals");
+  ASSERT_TRUE(inputSignal->connect());
+
+  sig_sm.update();
+  auto valueSignal = std::shared_ptr<const Signal>();
+  ReceiveStatus ret = inputSignal->receive(valueSignal, 500ms);
+  ASSERT_EQ(ReceiveStatus::RECEIVE_OK, ret);
+  EXPECT_EQ(valueSignal->value, 0);
```
Comment (gustavoneves12) [inline tests/test_signal_generator.cpp]: I would use `EXPECT_NEAR`

File: tests/test_signal_generator.cpp [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+  __assert_and_run_state_machine(sm, conf, endpoint)
+
+
+
+namespace hyro
+{
+namespace test
+{
+
+TEST (TestSignalGenerator, SignalGeneratorComponentCheck)
```
Comment (gustavoneves12) [inline tests/test_signal_generator.cpp:36]: Your tests are very well programmed. However, I would check the value of the signals output channel after changing the dynamic parameters.

File: tests/test_signal_generator.cpp [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  ASSERT_TRUE(dynamic_property_access.get("amplitude", amplitude));
+  ASSERT_TRUE(dynamic_property_access.get("frequency", frequency));
+  ASSERT_TRUE(dynamic_property_access.get("cosine", cosine));
+
+  ASSERT_NEAR(amplitude, 10.0, 0.001);
+  ASSERT_NEAR(frequency, 20.0, 0.001);
+  ASSERT_EQ(cosine,true);
+}
+
+TEST (TestSignalGenerator, DigitalConverterComponentCheck)
```
Comment (gustavoneves12) [inline tests/test_signal_generator.cpp:98]: Here you do not check the `digital_signals` output.

**Comment by murilomend** [comment]
@gustavoneves12  @joaobrittoneto  @geovanemimoso  
I just pushed the suggested modifications.

File: src/examples/signal_generator.cpp (hunk 10 above)
Comment (murilomend) [reply src/examples/signal_generator.cpp]: I decided to change like you suggested.

File: src/internal/SignalGenerator.cpp [hunk 17]
```diff
@@ -0,0 +1,56 @@
...
+
+
+SignalGenerator::~SignalGenerator()
+{
+
+}
+
+bool SignalGenerator::setAmplitude(const double amp)
+{
+  this->m_amp    = amp;
```
Comment (gustavoneves12) [inline src/internal/SignalGenerator.cpp:20]: @murilomend, I usually just include tabs like you did here, when I want to align with the operator **"="**.

I also avoid to use `this` inside the class. Only when it is required use `this`.

Please, replace for:

```
m_amp = amp;
```
Please, also apply these modifications in the other parts of the code.

**Comment by gustavoneves12** [comment]
@murilomend your code is very well programmed. :+1: 

Please, just fix the last suggestion, so that I can merge your code into master.

**Comment by gustavoneves12** [comment]
Tested! I am going to merged it.

File: src/components/SignalGeneratorComponent.cpp [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+hyro::Result SignalGeneratorComponent::start ()
+{
+  /** Start spinners and drivers (if any) here */
+  return hyro::Result::RESULT_OK;
+}
+
+hyro::Result SignalGeneratorComponent::update()
+{
+  Signal value;
+  value.timestamp = 1000;
```
Comment (joaobrittoneto) [inline src/components/SignalGeneratorComponent.cpp:109]: Timestamps are really important. 
Try to use the `timestamp = hyro::TimeUtils::Now_TimeSinceEpoch();`

Comment (gustavoneves12) [reply src/components/SignalGeneratorComponent.cpp:109]: My bad. 

@murilomend fix this and open a new MR.

Comment (murilomend) [reply src/components/SignalGeneratorComponent.cpp:109]: Sorry I just saw this review today. I will fix this and do another pull request.

**Pull request merged**